components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1169,64 +1165,27 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             //
             // If that, source peer still need to decide whether to destroy itself. When the target
             // peer has already moved on, source peer can destroy itself.
-            if epoch.get_version() > merge_target.get_region_epoch().get_version() {
+            if util::is_epoch_stale(merge_target.get_region_epoch(), r.get_region_epoch()) {
                 return Ok(true);
             }
             return Ok(false);
         }
+        drop(meta);
 
-        // Check whether target peer is set to tombstone already.
-        let state_key = keys::region_state_key(target_region_id);
-        if let Some(state) = self
-            .ctx
-            .engines
-            .kv
-            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)?
-        {
-            debug!(
-                "check target region local state";
-                "region_id" => self.region_id(),
-                "peer_id" => self.fsm.peer_id(),
-                "target_region_id" => target_region_id,
-                "state" => ?state,
-            );
-            if state.get_state() == PeerState::Tombstone
-                && state.get_region().get_region_epoch().get_conf_ver()
-                    >= merge_target.get_region_epoch().get_conf_ver()
-            {
-                // Replica was destroyed.
-                return Ok(true);
+        // All of the target peers must exist before merging which is guaranteed by PD.
+        // Now the target peer is not in region map, so if everything is ok, the local target
+        // region should be fresher than merge target
+        if self.is_local_merge_target_region_fresher(merge_target)? {
+            Ok(true)
+        } else {
+            if self.ctx.cfg.ensure_all_target_peer_exist {
+                debug_assert!(
```

Comment:
I think this is useless. It's optimized out in release mode, and we don't test debug build with PD, so the check can't really detect anything about *real* PD, it just checks for nothing.

Reply:
Maybe using something else instead. 
I think we need a method to assert in release mode in the development environment.
There are some cases that using `assert!` directly is a little bit overkill in the production environment but necessary in the development environment. Then if enough tests have been passed, maybe we can change it to `assert!`.

Reply:
I don't think it needs to be `assert` in any cases. Even this is not hold, it doesn't effect normal operations of TiKV, it just leads to some unremovable replicas. I think just an alert is enough and leave an API to allow user to clean up stale replicas.

Reply:
It's ok to not change to `assert`. but an assert method is needed to detect some bugs in our tests in the development environment.

Reply:
Yes, that's why I suggest an alert.